File: inference/pipeline.py
"""Contract -> Review: the three-task adapter run over one live document.

Contracts longer than the context budget are split into overlapping windows
(the cluster's CUAD path was windowed too, see macbundle/README). Windows are
the outer loop and questions the inner one, so each window's prefill is paid
once and reused for every question through the engine's prefix cache.

Output matches webapp/lib/types.ts `Review`.
"""

import sys
import uuid
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
sys.path.insert(0, str(ROOT / "macbundle" / "lib"))

import prompts as P  # noqa: E402
from doc_harness import parse_output, _find_json, strip_wrappers  # noqa: E402
from spans import norm_with_map, spans_of  # noqa: E402
from text_norm import norm  # noqa: E402

from engine import MAX_CONTEXT, MAX_NEW_TOKENS  # noqa: E402

# The system prompt the task-3 validation rows (and so the cluster's bf16
# output) actually use. prompts.TASK3_SYSTEM is the longer teacher prompt;
# on validation_60 it agreed with the cluster less (token Jaccard 0.39 vs 0.60).
TASK3_SYSTEM_TRAINED = (
    "You are a contract review assistant. Given a clause and its category, write "
    "EXACTLY ONE sentence for a business reader explaining what the clause does in "
    "practice and what brings it into effect."
)

# Room left in the prompt for the system message, the question and the
# chat-template tokens around them. The longest position is ~40 tokens.
PROMPT_OVERHEAD = 400
WINDOW_TOKENS = MAX_CONTEXT - MAX_NEW_TOKENS - PROMPT_OVERHEAD
OVERLAP_TOKENS = 1024
# a window edge moves back to the nearest line break within this many chars
SNAP_CHARS = 400

POSITION_STATUS = {
    "Entailment": ("Meets standard", "grn", "The contract supports this position."),
    "Contradiction": ("Needs attention", "red", "The contract appears to conflict with this position."),
    "NotMentioned": ("Not addressed", "amb", "The contract does not address this."),
}
# Across windows a conflict anywhere outranks support anywhere, which
# outranks silence: a clause only has to appear in one window to count.
VERDICT_RANK = {"Contradiction": 0, "Entailment": 1, "NotMentioned": 2}


def windows(tok, text):
    """-> [(start_char, end_char)] covering text, each within WINDOW_TOKENS."""
    enc = tok._tokenizer(text, add_special_tokens=False, return_offsets_mapping=True)
    offs = enc["offset_mapping"]
    if len(offs) <= WINDOW_TOKENS:
        return [(0, len(text))]
    out, t = [], 0
    while True:
        start = offs[t][0]
        last = t + WINDOW_TOKENS
        if last >= len(offs):
            out.append((start, len(text)))
            return out
        end = offs[last][0]
        nl = text.rfind("\n", end - SNAP_CHARS, end)
        if nl > start:
            end = nl + 1
        out.append((start, end))
        # next window starts OVERLAP_TOKENS before this one ends
        t = next(i for i, (a, _) in enumerate(offs) if a >= end) - OVERLAP_TOKENS
        nl = text.rfind("\n", offs[t][0] - SNAP_CHARS, offs[t][0])
        if nl > start:
            t = next(i for i, (a, _) in enumerate(offs) if a >= nl + 1)


def parse_cuad(raw):
    obj = _find_json(strip_wrappers(raw))
    if not isinstance(obj, dict) or not isinstance(obj.get("present"), bool):
        return None, []
    ev = obj.get("evidence") or []
    if isinstance(ev, str):
        ev = [ev]
    return obj["present"], [e for e in ev if isinstance(e, str) and e.strip()]


def _dedupe(texts):
    seen, out = set(), []
    for t in texts:
        k = norm(t)
        if k and k not in seen:
            seen.add(k)
            out.append(t)
    return out


class Reviewer:
    def __init__(self, engine):
        self.eng = engine

    def run(self, text, tasks=("compliance", "clauses"), document_name="Contract",
            positions=None, categories=None, progress=lambda **kw: None):
        positions = positions or P.PLAYBOOK_POSITIONS
        categories = categories or P.CLAUSE_CATEGORIES
        wins = windows(self.eng.tok, text)
        dn, di = norm_with_map(text)
        assert dn == norm(text), "norm_with_map drifted from norm"

        qs = []
        if "compliance" in tasks:
            qs += [("nli", q) for q in positions]
        if "clauses" in tasks:
            qs += [("cuad", c) for c in categories]
        total = len(wins) * len(qs)
        done, stats = 0, {"windows": len(wins), "unparsed": 0, "unlocated_quotes": 0}

        per_q = {q: [] for q in qs}  # (label, evidence) per window
        for w, (a, b) in enumerate(wins):
            chunk = text[a:b]
            self.eng.reset_cache()
            for kind, q in qs:
                if kind == "nli":
                    msgs = [{"role": "system", "content": P.TASK1_SYSTEM},
                            {"role": "user", "content": P.TASK1_USER.format(text=chunk, question=q)}]
                    raw = self.eng.generate(msgs, reuse_prefix=True).text
                    p = parse_output(raw)
                    label, ev = p["verdict"], p["evidence"]
                else:
                    msgs = [{"role": "system", "content": P.TASK2_SYSTEM},
                            {"role": "user", "content": P.TASK2_USER.format(text=chunk, category=q)}]
                    raw = self.eng.generate(msgs, reuse_prefix=True).text
                    label, ev = parse_cuad(raw)
                if label is None:
                    stats["unparsed"] += 1
                per_q[(kind, q)].append((label, ev))
                done += 1
                progress(stage="clauses" if kind == "cuad" else "positions",
                         done=done, total=total, window=w + 1, windows=len(wins))
        self.eng.reset_cache()

        def located(ev):
            s = spans_of(text, _dedupe(ev))
            stats["unlocated_quotes"] += len(_dedupe(ev)) - len(s)
            return s

        compliance, clauses = [], []
        for i, q in enumerate(positions if "compliance" in tasks else []):
            res = [r for r in per_q[("nli", q)] if r[0]]
            verdict = min((r[0] for r in res), key=VERDICT_RANK.get, default="NotMentioned")
            ev = [e for v, evs in res if v == verdict for e in evs]
            status, tone, note = POSITION_STATUS[verdict]
            compliance.append({"id": f"p{i}", "kind": "position", "title": q, "status": status,
                               "tone": tone, "evidence": located(ev), "note": note})

        found = []
        for i, c in enumerate(categories if "clauses" in tasks else []):
            res = per_q[("cuad", c)]
            present = any(r[0] for r in res)
            ev = [e for v, evs in res if v for e in evs]
            item = {"id": f"c{i}", "kind": "clause", "title": c,
                    "status": "Found" if present else "Not detected",
                    "tone": "grn" if present else "gry",
                    "evidence": located(ev) if present else [], "note": None}
            clauses.append(item)
            if present and item["evidence"]:
                found.append(item)

        for k, item in enumerate(found):
            progress(stage="notes", done=k, total=len(found))
            clause = " ".join(s["text"] for s in item["evidence"])
            msgs = [{"role": "system", "content": TASK3_SYSTEM_TRAINED},
                    {"role": "user", "content": P.TASK3_USER.format(cat=item["title"], clause=clause)}]
            item["note"] = strip_wrappers(self.eng.generate(msgs).text).strip()

        return {
            "id": uuid.uuid4().hex[:10],
            "documentName": document_name,
            "representing": "Receiving Party" if "compliance" in tasks else "Counterparty",
            "counterparty": "Disclosing Party" if "compliance" in tasks else "—",
            "playbookName": "Standard playbook",
            "documentText": text,
            "compliance": compliance,
            "clauses": clauses,
            "stats": stats,
        }

File: webapp/lib/server/pipeline.ts
/**
 * Contract -> Review, against the served model. Port of inference/pipeline.py.
 *
 * Questions run concurrently: the server answers many at once and caches the
 * shared contract prefix itself, which is what turns a 230 s local review into
 * ~17 s. Windows are still the outer loop so a long contract is covered.
 */

import { ask, countTokens, PROMPT_BUDGET, type Msg } from "./hf";
import {
  CLAUSE_CATEGORIES, PLAYBOOK_POSITIONS, POSITION_STATUS, SILENT_NOTE,
  TASK1_SYSTEM, TASK2_SYSTEM, TASK3_SYSTEM, task1User, task2User, task3User,
} from "./prompts";
import { norm, spansOf } from "./spans";
import type { Item, Progress, Task, Tone } from "../types";

const PARALLEL = 8;
// leaves room for the system message, the question and the template tokens
const PROMPT_OVERHEAD = 400;
const WINDOW_TOKENS = PROMPT_BUDGET - PROMPT_OVERHEAD;
const OVERLAP_CHARS = 4000;
const SNAP_CHARS = 400;
// conservative: contracts run ~3.6 chars/token, so this over-estimates tokens
const CHARS_PER_TOKEN = 3.2;

// Across windows a conflict anywhere outranks support anywhere, which outranks
// silence; a clause only has to appear in one window to count.
const VERDICT_RANK: Record<string, number> = { Contradiction: 0, Entailment: 1, NotMentioned: 2 };

/** Run fn over items with at most `limit` in flight, results in order. */
async function pool<T, R>(items: T[], limit: number, fn: (t: T, i: number) => Promise<R>): Promise<R[]> {
  const out = new Array<R>(items.length);
  let next = 0;
  await Promise.all(Array.from({ length: Math.min(limit, items.length) }, async () => {
    for (let i = next++; i < items.length; i = next++) out[i] = await fn(items[i], i);
  }));
  return out;
}

function stripWrappers(raw: string): string {
  return (raw || "")
    .replace(/<think>[\s\S]*?<\/think>/gi, " ")
    .replace(/^[\s\S]*?<\/think>/i, (m) => (/<\/think>/i.test(m) ? " " : m))
    .replace(/```(?:json)?/gi, " ")
    .trim();
}

/** First balanced JSON object in the text, like doc_harness._find_json. */
function findJson(text: string): Record<string, unknown> | null {
  try {
    const o = JSON.parse(text);
    if (o && typeof o === "object") return o as Record<string, unknown>;
  } catch { /* fall through to the brace scan */ }
  let depth = 0, start = -1;
  for (let i = 0; i < text.length; i++) {
    if (text[i] === "{") { if (depth === 0) start = i; depth++; }
    else if (text[i] === "}") {
      depth--;
      if (depth === 0 && start >= 0) {
        try {
          const o = JSON.parse(text.slice(start, i + 1));
          if (o && typeof o === "object") return o as Record<string, unknown>;
        } catch { start = -1; }
      }
    }
  }
  return null;
}

const evidenceOf = (o: Record<string, unknown> | null): string[] => {
  const ev = o?.evidence;
  const list = typeof ev === "string" ? [ev] : Array.isArray(ev) ? ev : [];
  return list.filter((e): e is string => typeof e === "string" && e.trim().length > 0);
};

const VERDICTS = ["Entailment", "Contradiction", "NotMentioned"];

function parseVerdict(raw: string): [string | null, string[]] {
  const o = findJson(stripWrappers(raw));
  const v = typeof o?.verdict === "string"
    ? VERDICTS.find((c) => c.toLowerCase() === (o.verdict as string).replace(/[\s_-]/g, "").toLowerCase())
    : undefined;
  return [v ?? null, evidenceOf(o)];
}

function parsePresent(raw: string): [boolean | null, string[]] {
  const o = findJson(stripWrappers(raw));
  return [typeof o?.present === "boolean" ? o.present : null, evidenceOf(o)];
}

/** Shrink [start,end) until the server's tokenizer says it fits the window. */
async function fit(text: string, start: number, end: number): Promise<number> {
  for (let i = 0; i < 6; i++) {
    const n = await countTokens(text.slice(start, end));
    if (n === null || n <= WINDOW_TOKENS) return end;
    // scale by the measured ratio, with a little margin, then snap to a line
    const target = Math.floor((end - start) * (WINDOW_TOKENS / n) * 0.95);
    const shrunk = start + Math.max(target, 1000);
    if (shrunk >= end) return end;
    const nl = text.lastIndexOf("\n", shrunk);
    end = nl > start + SNAP_CHARS ? nl + 1 : shrunk;
  }
  return end;
}

/**
 * Character ranges covering the text, each within the model's window.
 *
 * The character estimate only picks the starting size: a token-dense contract
 * can still exceed the window, and the endpoint rejects the whole request when
 * it does, so every window is checked against the real tokenizer.
 */
export async function windows(text: string): Promise<[number, number][]> {
  if (text.length / CHARS_PER_TOKEN <= WINDOW_TOKENS) {
    const n = await countTokens(text);
    if (n === null || n <= WINDOW_TOKENS) return [[0, text.length]];
  }
  const span = Math.floor(WINDOW_TOKENS * CHARS_PER_TOKEN);
  const out: [number, number][] = [];
  let start = 0;
  while (start < text.length) {
    let end = Math.min(start + span, text.length);
    if (end < text.length) {
      const nl = text.lastIndexOf("\n", end);
      if (nl > start + SNAP_CHARS) end = nl + 1;
    }
    end = await fit(text, start, end);
    out.push([start, end]);
    if (end >= text.length) break;
    start = Math.max(end - OVERLAP_CHARS, start + 1);
  }
  return out;
}

export interface RunOptions {
  tasks: Task[];
  positions?: string[];          // an uploaded playbook, else the standard one
  onProgress?: (p: Progress) => void | Promise<void>;
}

export interface RunResult {
  compliance: Item[];
  clauses: Item[];
  stats: { windows: number; unparsed: number; unlocated_quotes: number };
  representing: string;
  counterparty: string;
  playbookName: string;
}

export async function runReview(text: string, { tasks, positions, onProgress }: RunOptions): Promise<RunResult> {
  const playbook = positions?.length ? positions : PLAYBOOK_POSITIONS;
  const wins = await windows(text);
  const qs: { kind: "nli" | "cuad"; q: string }[] = [
    ...(tasks.includes("compliance") ? playbook.map((q) => ({ kind: "nli" as const, q })) : []),
    ...(tasks.includes("clauses") ? CLAUSE_CATEGORIES.map((q) => ({ kind: "cuad" as const, q })) : []),
  ];
  const stats = { windows: wins.length, unparsed: 0, unlocated_quotes: 0 };
  const total = wins.length * qs.length;
  let done = 0;

  const perQ = new Map<string, [string | boolean | null, string[]][]>();
  for (const [w, [a, b]] of wins.entries()) {
    const chunk = text.slice(a, b);
    const answers = await pool(qs, PARALLEL, async ({ kind, q }) => {
      const msgs: Msg[] = kind === "nli"
        ? [{ role: "system", content: TASK1_SYSTEM }, { role: "user", content: task1User(chunk, q) }]
        : [{ role: "system", content: TASK2_SYSTEM }, { role: "user", content: task2User(chunk, q) }];
      const raw = await ask(msgs);
      const parsed = kind === "nli" ? parseVerdict(raw) : parsePresent(raw);
      if (parsed[0] === null) stats.unparsed++;
      done++;
      if (done % 4 === 0 || done === total) {
        await onProgress?.({ stage: kind === "cuad" ? "clauses" : "positions", done, total,
                             window: w + 1, windows: wins.length });
      }
      return parsed;
    });
    qs.forEach(({ kind, q }, i) => {
      const key = `${kind}:${q}`;
      perQ.set(key, [...(perQ.get(key) ?? []), answers[i]]);
    });
  }

  const located = (ev: string[]) => {
    const { spans, dropped } = spansOf(text, ev);
    stats.unlocated_quotes += dropped;
    return spans;
  };

  const compliance: Item[] = [];
  if (tasks.includes("compliance")) {
    playbook.forEach((q, i) => {
      const res = (perQ.get(`nli:${q}`) ?? []).filter(([v]) => v) as [string, string[]][];
      const verdict = res.map(([v]) => v).sort((a, b) => VERDICT_RANK[a] - VERDICT_RANK[b])[0] ?? "NotMentioned";
      const ev = res.filter(([v]) => v === verdict).flatMap(([, e]) => e);
      const [status, tone, assessment] = POSITION_STATUS[verdict as keyof typeof POSITION_STATUS];
      const silent = verdict === "NotMentioned";
      compliance.push({
        id: `p${i}`, kind: "position", title: q, status, tone: tone as Tone,
        evidence: located(ev), assessment,
        note: silent ? SILENT_NOTE : null, noteSource: silent ? "fixed" : null,
      });
    });
  }

  const clauses: Item[] = [];
  if (tasks.includes("clauses")) {
    CLAUSE_CATEGORIES.forEach((c, i) => {
      const res = perQ.get(`cuad:${c}`) ?? [];
      const present = res.some(([p]) => p === true);
      const ev = res.filter(([p]) => p === true).flatMap(([, e]) => e);
      clauses.push({
        id: `c${i}`, kind: "clause", title: c,
        status: present ? "Found" : "Not detected", tone: present ? "grn" : "gry",
        evidence: present ? located(ev) : [], note: null, noteSource: null,
      });
    });
  }

  return {
    compliance, clauses, stats,
    representing: tasks.includes("compliance") ? "Receiving Party" : "Counterparty",
    counterparty: tasks.includes("compliance") ? "Disclosing Party" : "—",
    playbookName: positions?.length ? "Uploaded playbook" : "Standard playbook",
  };
}

export { norm };

/**
 * One finding's "why it matters" note, written on request.
 *
 * Task 3 was trained on (clause category, clause text), so it needs the quoted
 * evidence: a finding with nothing quoted has nothing to explain.
 */
export async function writeNote(title: string, evidence: string[]): Promise<string> {
  if (!evidence.length) throw new Error("This finding has no supporting language to explain.");
  const raw = await ask([{ role: "system", content: TASK3_SYSTEM },
                         { role: "user", content: task3User(title, evidence.join(" ")) }]);
  return stripWrappers(raw).trim();
}
